src/NuGet.Clients/NuGet.SolutionRestoreManager/VulnerablePackagesInfoBar.cs

```diff
@@ -0,0 +1,163 @@
+        {
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            if (!_visible)
+            {
+                return;
+            }
+
+            _closeFromHide = true;
+
+            try
+            {
+                if (_eventCookie.HasValue)
+                {
+                    _infoBarUIElement?.Unadvise(_eventCookie.Value);
+                    _infoBarUIElement?.Close();
+                    _infoBarUIElement = null;
+                }
+            }
+            finally
+            {
+                _closeFromHide = false;
+            }
+        }
+
+        protected InfoBarModel GetInfoBarModel()
+        {
+            return new InfoBarModel(
+                Resources.InfoBar_TextMessage,
+                new IVsInfoBarActionItem[] {
+                    new InfoBarHyperlink(Resources.InfoBar_HyperlinkMessage),
+                },
+                KnownMonikers.StatusWarning);
+        }
+
+        internal async Task CreateAndShowInfoBarAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            if (_closed || _visible)
+            {
+                return;
+            }
+
+            IVsInfoBarHost? infoBarHost;
+            try
+            {
+                IVsUIShell? uiShell = await _asyncServiceProvider.GetServiceAsync<SVsUIShell, IVsUIShell>(throwOnFailure: false);
+                if (uiShell == null)
+                {
+                    throw new NullReferenceException(nameof(uiShell));
+                }
+
+                if (ErrorHandler.Failed(uiShell.FindToolWindow((uint)__VSFINDTOOLWIN.FTW_fFindFirst, VSConstants.StandardToolWindows.SolutionExplorer, out var windowFrame)))
+                {
+                    throw new NullReferenceException(nameof(windowFrame));
+                }
+
+                object tempObject;
+                if (ErrorHandler.Failed(windowFrame.GetProperty((int)__VSFPROPID7.VSFPROPID_InfoBarHost, out tempObject)))
+                {
+                    throw new NullReferenceException(nameof(tempObject));
+                }
+
+                infoBarHost = (IVsInfoBarHost)tempObject;
+            }
+            catch (Exception ex)
+            {
+                await TelemetryUtility.PostFaultAsync(ex, nameof(VulnerablePackagesInfoBar));
+                return;
+            }
+
+            try
+            {
+                IVsInfoBarUIFactory? infoBarFactory = await _asyncServiceProvider.GetServiceAsync<SVsInfoBarUIFactory, IVsInfoBarUIFactory>(throwOnFailure: false);
+                if (infoBarFactory == null)
+                {
+                    NullReferenceException exception = new NullReferenceException(nameof(infoBarFactory));
+                    await TelemetryUtility.PostFaultAsync(exception, nameof(VulnerablePackagesInfoBar));
```

Comment:
I wonder if that telemetry would be actionable though without the information added with the Exception.

Reply:
Telemetry's FaultEvent class requires an `Exception` object on all constuctors: https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.telemetry.faultevent.-ctor?view=visualstudiosdk-2022 (probably because there are some properties that get populated from the exception itself, such as the exception type).

If we don't want to create exceptions, then we can't use fault telemetry, but we can use "normal" telemetry. But then we can't use prism to automatically collect faults, rank them, etc.